core/command_parser.h

```diff
@@ -0,0 +1,246 @@
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef COMMAND_PARSER_H
+#define COMMAND_PARSER_H
+
+#include "core/object/reference.h"
+#include "core/templates/ordered_hash_map.h"
+
+class CommandOption : public Reference {
+	GDCLASS(CommandOption, Reference);
+
+	struct ArgumentChecker;
+	struct FunctionChecker;
+	struct CallableChecker;
+
+	// Names for the command. e.g --help or -h.
+	PackedStringArray _names;
+	// List of allowed values for each argument, empty if any value is allowed.
+	PackedStringArray _default_args;
+	// List of allowed values for each argument, empty if any value is allowed.
+	PackedStringArray _allowed_values;
+	// Command description that will be displayed in help.
+	String _description;
+	// Command category, commands sharing the same category are grouped together in the help text.
+	String _category;
+	// Name for the command arguments that will be displayed in help text.
+	String _metavar = RTR("<arg>");
+	// Make the command visible in help.
+	bool _show_in_help = true;
+	// If true, arguments can be specified without option name.
+	bool _positional = false;
+	// If true, argument always should be provided.
+	bool _required = false;
+	// If true, the command can be specified several times.
+	bool _multitoken = false;
+	// Arguments count required for the command, -1 for all arguments left.
+	int _args_count = 0;
+	// Checker for each argument, nullptr if not check specified.
+	ArgumentChecker *_checker = nullptr;
+
+protected:
+	static void _bind_methods();
+
+public:
+	using CheckFunction = bool (*)(const String &);
+
+	void set_names(const PackedStringArray &p_names);
+	PackedStringArray get_names() const;
+	void set_default_args(const PackedStringArray &p_args);
+	PackedStringArray get_default_args() const;
+	void set_allowed_values(const PackedStringArray &p_values);
+	PackedStringArray get_allowed_values() const;
+	void set_description(const String &p_description);
+	String get_description() const;
+	void set_category(const String &p_category);
+	String get_category() const;
+	void set_metavar(const String &p_metavar);
+	String get_metavar() const;
+	void set_args_count(int p_count);
+	int get_args_count() const;
+	void set_show_in_help(const bool p_enabled);
+	bool is_show_in_help() const;
+	void set_positional(bool p_positional);
+	bool is_positional() const;
+	void set_required(bool p_required);
+	bool is_required() const;
+	void set_multitoken(bool p_multitoken);
+	bool is_multitoken() const;
+
+	void set_static_checker(CheckFunction p_function, const String &p_error_msg);
+	Error set_checker(const Callable &p_callable, const String &p_error_msg);
+	const ArgumentChecker *get_checker() const;
+	void remove_checker();
+
+	void set_data(const String &p_category, const String &p_description);
```

Comment:
Which methods do you suggest to export?

Reply:
From most commonly used to less commonly used.

category, description, arg count, default values, allowed values.

Exposing boolean parameters here is probably not worth it because GDScript does not allow to use keyword arguments to be explicit about it.

Reply:
Sounds good, will do.

Reply:
I realized that we have such options in constructor. Maybe just rename this function? How about `set_help_info`?

Reply:
Constructors are not available in GDScript, though. `set_help_info()` will make sense if we keep category/description parameters only. But to be honest, not sure if having this convenience function is worth it if it would only save 2 lines of code. It would be a lot of useful in GDScript if this method could provide initialization of other properties as suggested above.

Reply:
> But to be honest, not sure if having this convenience function is worth it if it would only save 2 lines of code.

Makes sense. Maybe just remove this function at all? Or you would prefer to have such function with extended number of parameters?

Reply:
I'd remove the function and see if it's worth implementing in the future, according to use cases. 🙂

Reply:
Okay, I remove it then :)